lc0/src/neural/network_cudnn.cu

```diff
@@ -0,0 +1,836 @@
+
+    BaseLayer(int c, int h, int w, BaseLayer *ip);
+
+    // input2 is optional (skip connection)
+    size_t getOutputSize() { return bpe * N * C * H * W; }
+    virtual void eval(void *output, void *input, void *input2, void *scratch) = 0;
+};
+
+class ConvLayer : public BaseLayer
+{
+private:
+    int Cinput;
+    int filterSize;
+
+    void *biases;
+    void *weights;
+
+    bool useRelu;
+    bool useBias;
+
+    cudnnFilterDescriptor_t			filterDesc;
+    cudnnConvolutionDescriptor_t	convDesc;
+    cudnnConvolutionFwdAlgo_t       convAlgo;
+
+    cudnnTensorDescriptor_t         biasDesc;
+    cudnnTensorDescriptor_t         inTensorDesc;
+    cudnnTensorDescriptor_t         outTensorDesc;
+    cudnnActivationDescriptor_t     activation;
+
+
+public:
+    ConvLayer(BaseLayer *ip, int C, int H, int W, int size, int Cin, bool relu = false, bool bias = false);
+    ~ConvLayer();
+    void loadWeights(void *pfilter, void *pBias = NULL);
+    void eval(void *output, void *input, void *input2, void *scratch) override;
+};
+
+class SoftMaxLayer : public BaseLayer
+{
+private:
+    cudnnTensorDescriptor_t         outTensorDesc;
+
+public:
+    SoftMaxLayer(BaseLayer *ip);
+    void eval(void *output, void *input, void *input2, void *scratch) override;
+};
+
+class BNLayer : public BaseLayer
+{
+private:
+    void *means;
+    void *variances;
+    bool useRelu;
+
+public:
+    BNLayer(BaseLayer *ip, bool relu);
+    ~BNLayer();
+
+    void loadWeights(void *cpuMeans, void *cpuVar);
+    void eval(void *output, void *input, void *input2, void *scratch) override;
+
+};
+
+class FCLayer : public BaseLayer
+{
+private:
+    void *weights;
+    void *biases;
+    bool useBias;
+    bool useRelu;
+    bool useTanh;
+
+public:
+    FCLayer(BaseLayer *ip, int C, int H, int W, bool relu, bool bias, bool tanh = false);
+    ~FCLayer();
+
+    void loadWeights(void *cpuWeight, void *cpuBias);
+    void eval(void *output, void *input, void *input2, void *scratch) override;
+};
+
+
+// Each residual block has (4 kernels per block)
+// A convolution of 128 filters of kernel size 3 � 3 with stride 1
+
+// Batch normalisation
+// A rectifier non - linearity
+
+// A convolution of 128 filters of kernel size 3 � 3 with stride 1
+
+// Batch normalisation
+// A skip connection that adds the input to the block
+// A rectifier non - linearity
+
+// need implementations of 
+//  1. convolution layer (no bias/activation needed)
+//  2. Fully connected layer (with optional bias, and optional relu),
+//  3. batch normilization with optional sum (skip connection) and RELU
+
+// Need memory for 3 data buffers
+//  1. input for the layer
+//  2. output of the layer
+//  3. data from old layer for skip connection
+
+
+/////////////////////////////////////////////////////////////////////////////
+//                      Global/Static variables                            //
+/////////////////////////////////////////////////////////////////////////////
+
+int BaseLayer::N = 1;
```

Comment:
I believe that if you make them constexpr, you can initialize them right in a class definition.

Reply:
Then it should not be static. We'll need to support running it on multiple GPUs soon, and having it static will make it harder.

Reply:
fixed. Made it a parameter to eval.